Add modal refactor (V3?)

hahnbeelee opened the pull request:
### Summary
- Made the add modal look more like [Ein's newest designs](https://www.figma.com/file/81R3CvQY1EgxFZSxtEJNrj/Ein-Working-File?node-id=540%3A407)
- Created a dropdown for non-auto-fulfilled classes
- Added empty state
- When you select a class, the class name is no longer an editable input (last time i tried to do this though it caused some errors with the input which is why it wasn't implemented... it's looking like it's acting normal as of now, but i haven't extensively tested)

### Screenshots / Flow
![image](https://user-images.githubusercontent.com/55263191/110219932-3553b200-7e90-11eb-95a0-8b7303f643dc.png)
![image](https://user-images.githubusercontent.com/55263191/110219938-41d80a80-7e90-11eb-9bba-b006a9680234.png)
![image](https://user-images.githubusercontent.com/55263191/110219945-4bfa0900-7e90-11eb-910d-fc0e0d658900.png)
![image](https://user-images.githubusercontent.com/55263191/110219948-52888080-7e90-11eb-8302-fab8dfdd9da5.png)

Empty state:
![image](https://user-images.githubusercontent.com/55263191/110220035-ff62fd80-7e90-11eb-9a87-fccc93b4f918.png)

### Test Plan
- Add classes + make sure the requirements side-bar fills up requirements accordingly!

### Notes
Differences between designs @einc 
- I added an option for adding no potential/related requirements in case the user doesn't want to add a class for any of them
- After you add a potential/related requirement it becomes bold so that it affirms the user it selected the requirement.

SamChou19815 commented on src/components/Modals/NewCourse/RequirementsDropdown.vue:
```diff
@@ -0,0 +1,171 @@
+        <img class="warning-icon" src="@/assets/images/warning.svg" alt="warning-icon" />
+        {{ selected }}
+      </div>
+      <drop-down-arrow :isFlipped="showDropdown" :fillColor="'#148481'" />
```
new constant?

SamChou19815 commented on src/components/Modals/NewCourse/RequirementsDropdown.vue:
```diff
@@ -0,0 +1,171 @@
+        &:active,
+        &:focus,
+        &:hover {
+          background: rgba(50, 160, 242, 0.1);
```
new constant?

SamChou19815 reviewed (commented):
Some nits on color constants.

benjamin-shen commented on src/components/Modals/NewCourse/SelectedRequirementEditor.vue:
```diff
@@ -88,15 +65,31 @@ export default Vue.extend({
+      return this.relatedRequirements.concat(this.potentialRequirements);
+    },
+    nonAutoRequirementsText(): string {
+      if (this.selectedRequirementID === '') this.nonAutoRequirements.map(it => it.name).join(', ');
```
Do you need to `return`?

benjamin-shen commented on src/components/Modals/NewCourse/RequirementsDropdown.vue:
```diff
@@ -0,0 +1,171 @@
+      :aria-expanded="showDropdown"
+    >
+      {{ selected }}
+      <drop-down-arrow :isFlipped="showDropdown" :fillColor="'#148481'" />
```
As Sam said, constant here?

benjamin-shen commented on src/components/Modals/NewCourse/SelectedRequirementEditor.vue:
```diff
@@ -1,75 +1,52 @@
     </div>
-    <div v-if="relatedRequirements.length > 0">
+    <div v-else class="newCourse-requirements-container newCourse-requirements">
+      This class does not automatically fulfill any requirements
```
Is this text supposed to be blue? Also, should there be a period?

hahnbeelee replied:
i realized @einc doesnt use periods lol. i agree that there should be a period. Ein, is it cool if i add a period or is this a design decision?

einc replied:
no lol that's just my bad typing habits. yes haha pls add periods or capitalize stuff as normal <3

benjamin-shen reviewed (commented):
Looks good! I left some comments.

A few more high-level comments:
- I'm personally not in love with the empty dropdown text, especially after choosing None of the above. (maybe designers have a better suggestion?)
- Can the dropdown be closed when you click outside of it?
- The dropdown options width is a bit wonky on mobile. This is probably lower priority, but it'd be sick if you could fix it.

einc commented:
> * I'm personally not in love with the empty dropdown text, especially after choosing None of the above. (maybe designers have a better suggestion?)

Can we have the placeholder dropdown be "Select Requirement (optional)" looking like this:
(and get rid of the None option?)
<img width="725" alt="Screen Shot 2021-03-07 at 12 27 41 AM" src="https://user-images.githubusercontent.com/14082828/110230048-f77c7b00-7edb-11eb-9679-9a773a2b09b3.png">

einc commented:
<img width="952" alt="Screen Shot 2021-03-07 at 12 33 03 AM" src="https://user-images.githubusercontent.com/14082828/110230127-b33daa80-7edc-11eb-893c-9cabbe17a2bf.png">
I'm a bit confused why the Computer Science Core requirement isn't automatically fulfilled here? [this is CS major in CoE]

einc commented:
sorry, I should have changed this in the designs before, but could we also change the "We cannot check the requirements marked with a warning sign, so double check before selecting" text to "**We cannot accurately check the requirements marked with the warning icon, so double check before selecting."**

einc commented:
HAHNBEE!! THIS IS 🔥!!! left some small comments but looks pretty good overall!
woohoo on also getting the class name to be non-editable 🥳
thanks for helping getting the add modal right and making it make sense!!

hahnbeelee commented:
> > * I'm personally not in love with the empty dropdown text, especially after choosing None of the above. (maybe designers have a better suggestion?)
> 
> Can we have the placeholder dropdown be "Select Requirement (optional)" looking like this:
> (and get rid of the None option?)
> <img alt="Screen Shot 2021-03-07 at 12 27 41 AM" width="725" src="https://user-images.githubusercontent.com/14082828/110230048-f77c7b00-7edb-11eb-9679-9a773a2b09b3.png">

If we don't have a 'None' option there's no way for the user to unselect a requirement. So I feel like we should keep the None option to let the user revert their choice if they want to?

hahnbeelee commented:
> <img alt="Screen Shot 2021-03-07 at 12 33 03 AM" width="952" src="https://user-images.githubusercontent.com/14082828/110230127-b33daa80-7edc-11eb-893c-9cabbe17a2bf.png">
> 
> I'm a bit confused why the Computer Science Core requirement isn't automatically fulfilled here? [this is CS major in CoE] It's showing up in the dropdown but isn't self check?

hm yeah idk why that's happening. i feel like it isn't me though and is most likely the requirements calculations happening in the backend? I'll take a look at it after this PR.

SamChou19815 commented:
> <img alt="Screen Shot 2021-03-07 at 12 33 03 AM" width="952" src="https://user-images.githubusercontent.com/14082828/110230127-b33daa80-7edc-11eb-893c-9cabbe17a2bf.png">
> 
> I'm a bit confused why the Computer Science Core requirement isn't automatically fulfilled here? [this is CS major in CoE] It's showing up in the dropdown but isn't self check?

Because it's not marked as allow double counting yet

hahnbeelee commented:
Most recent commit:
- Added scss variable
- Made it look better on mobile
- Click outside closes the dropdown
- Changed the warning text
- Added placeholder text vvv
![image](https://user-images.githubusercontent.com/55263191/110248377-ba95a000-7f3e-11eb-8189-5348f07cfb54.png)

SamChou19815 commented on src/components/Modals/NewCourse/RequirementsDropdown.vue:
```diff
@@ -0,0 +1,188 @@
+
+import { clickOutside } from '@/utilities';
+import DropDownArrow from '@/components/DropDownArrow.vue';
+import { emGreen } from '@/assets/scss/_variables.scss';
```
to resolve the type error, add it to the type declaration in `src/assets/scss/_variables.scss.d.ts`

tcho6319 commented on src/components/Modals/NewCourse/SelectedRequirementEditor.vue:
```diff
@@ -1,75 +1,52 @@
-          {{ relatedRequirements.map(it => it.name).join(', ') }}
+      <div v-if="!editMode">
+        <div class="newCourse-requirements-container">
+          <!-- eslint-disable-next-line vue/no-v-html  -->
+          <div class="newCourse-title" v-html="nonAutoRequirementsText"></div>
```
Can you use a different approach instead of a `v-html` directive? As addressed in #298, v-html directives could possibly lead to a [security issue](https://vuejs.org/v2/guide/security.html#Injecting-HTML) with HTML injection

hahnbeelee replied:
I actually can't think of any other way to do this unfortunately :/ . Since, I'm logically rendering the emboldened text based on data.

tcho6319 replied:
@hahnbeelee Just pushed my commit for the implementation of an alternative solution. Let me know if you're ok with this. If you are, then this PR should be g2g!

tcho6319 commented on src/components/Modals/NewCourse/SelectedRequirementEditor.vue:
```diff
@@ -1,75 +1,52 @@
-    <div v-if="!editMode" class="newCourse-link" @click="toggleEditMode()">
-      {{ editRequirementsText }}
-    </div>
+    <div v-if="!editMode" class="newCourse-link" @click="toggleEditMode()">Edit Requirements</div>
```
Can we make this a button? The Add Modal is actually almost all keyboard accessible 🥰 except for this part!

tcho6319 reviewed (changes requested):
This looks great Hahnbee! Left a few suggestions in addition to the one below:
1. Can we document what `relatedRequirements` and `potentialRequirements` are in `NewCourseModal.vue` and what `nonAutoRequirements` are in `SelectedRequirementEditor.vue`?

tcho6319 commented on src/components/Modals/NewCourse/SelectedRequirementEditor.vue:
```diff
@@ -1,102 +1,115 @@
+      <div v-if="!editMode">
+        <div class="newCourse-requirements-container">
+          <div class="newCourse-title">
+            <span v-for="(reqName, index) in nonAutoRequirementsTextArray" :key="index">
+              <strong v-if="reqName.selected" class="newCourse-name">{{
+                index === nonAutoRequirementsTextArray.length - 1
+                  ? reqName.name
+                  : reqName.name + ', '
+              }}</strong>
+              <span v-else>{{
+                index === nonAutoRequirementsTextArray.length - 1
+                  ? reqName.name
+                  : reqName.name + ', '
+              }}</span>
+            </span>
```
@hahnbeelee I replaced `nonAutoRequirementsText` with `nonAutoRequirementsTextArray` which is an array of objects. Each object has the name of the requirement and a boolean for whether or not it was selected (and should have the strong tag). Using those fields, I conditionally apply the strong tag or span tag. I take a look at the index of `reqName` in `nonAutoRequirementsTextArray` to see if we need to have a `', '` or not.

hahnbeelee replied:
Looks good to me!
